### PR TITLE
Training overhaul- gpu, dataloaders, early stopping, seed tracking, torch nn LSTM

### PR DESCRIPTION
**Pull request by jdiaz4302**
Lots of significant changes that are not necessarily a lot of code lines, here is the list and rationale

- Filepaths were altered to agree with minimal usage of a Tallgrass interactive GPU instance; all the modeling outputs are now being stored on Tallgrass
- Lots of new config values, notably:
  - **Random seed** is now explicitly set at the beginning and is additionally used to modify file paths (`cell block 5`) so that we can later run some easily tracked experiments with different random initializations.
  - **Model hidden size** was increased and a small **dropout value** was used. The hidden size is intended to mimic an average model used within the division, and the dropout value was arbitrarily chosen to match the default value used in torch transformer modules (and provide some overfitting protection)
  - Values/settings that feed into the data loaders used for mini batches with subsequent larger models. This notebook is still a relatively small LSTM that can use the entire training partition at once.
  - **Early stopping patience** is now used (in `cell block 18`) to prevent overfitting and save time against unneeded epochs. This implementation/design was minimally modified Andy's lake repo. Early trial-and-error was used to determine 50 which I think is generally on the larger side, allowing more waste at the hopeful chance of a later improvement.
- Transition away from the LSTM implementation that I made `river-dl` users (which maintained tensorflow's _recurrent_ dropout option). If concerned, see https://github.com/jdiaz4302/lake_ice_prediction/pull/8 for some reassurance that predictive performance is very similar between the two, but I'm going to use some much larger models and would greatly benefit from the computation time benefits. See `cell block 13`. 
  - _With all of these changes implemented, performance has, again, incrementally improved in all catagories._
- Also, I'm tracking the duration each training loop now.
- The prediction generation and saving has some efficiency/memory implementations now to handle the larger models and GPU RAM bottlenecks.

# Requested review

Look over these training changes for any expected problems or sneaky bugs. I've demo'd this work on many, much larger models for good results, so it looks trustworthy, but this review will help ensure the quality/finality prior to running experiments with multiple random initializations of large models.

### CODE REVIEW

**Comment by jdiaz4302** [comment]
Note - I didn't include the new results or update the eval notebooks because this is a relay point before a much larger evaluation. But results did incrementally improve, and those can obviously be provided.

**Review by AndyMcAliley — APPROVED** [review]
This looks good! From a read-through, I didn't catch any bugs.

Have we discussed Jupytext before? It's an option worth considering if you anticipate submitting a lot of notebook-based PRs because the diffs that result should be cleaner.

File: 02_train/src/A_FitBasicLSTM.ipynb [hunk 1]
```diff
@@ -256,227 +244,311 @@
...
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
    "source": [
     "# initialize the model with a seed\n",
-    "torch.manual_seed(0)\n",
+    "torch.manual_seed(random_seed)\n",
+    "# initialize random mini batches with numpy seed\n",
+    "np.random.seed(random_seed)\n",
```
[the remark this answers is not in the document]
Comment (jdiaz4302) [reply 02_train/src/A_FitBasicLSTM.ipynb:321]: Yeah, I did notice CPU and GPU were slightly different, but I only planned on running them on GPU, so I wasn't concerned. Thanks for the link though, I'll document/note that at least.

**Comment by jdiaz4302** [comment]
> Have we discussed Jupytext before? It's an option worth considering if you anticipate submitting a lot of notebook-based PRs because the diffs that result should be cleaner.

Yeah, I'll look into that for the next PR because these diffs are awful
